### ISSUE TITLE
Catch janky/jittery animations that run at 60 fps

### ISSUE THREAD

**Issue by liyuqian** [flutter/flutter#19222]
A recent regression https://github.com/flutter/flutter/issues/813#issuecomment-403946266 shows that we can get jittery animations even if every frame is rendered fast. Some known causes include:
- Flutter animation framework generated a non-smooth sequence of frames (e.g., [an image suddenly disappeared in one frame](https://drive.google.com/open?id=1Y9tETl-Elr8_VAlapTTPqoqGHMJYmo75)).
- The input event is not handled smoothly due to irregular events delivery or VSYNC and input frequency mismatch (https://github.com/flutter/flutter/issues/41118).

It would be nice to set up a high-speed camera (i.e., some recent high-end smart phones) in our device lab, record the animations frame by frame, and compute some metrics to see how smooth transitions are during an animation.

**Comment by liyuqian** [comment]
@cbracken @Hixie : this might require a lot of work. But it seems to be the closest simulation of how our users actually see the animation, and it can catch some surprising issues that our current performance metrics may miss.

**Comment by Chozzle** [comment]
Ive noticed that transition animations using MaterialPageRoute are rarely 60fps on older devices. I've tested on an iPhone 6S and Samsung Note 4 (not particularly slow phones). The Flutter Gallery app suffers this problem, but even just adding one blank page to the demo app has this problem.

**Comment by Chozzle** [comment]
I don't see many issues here that prioritise performance improvements. I work for a large app development company that builds premium apps and this is by far the number one issue holding us back from using Flutter! Are the team aware of this? The comments above indicate it is a bit of a surprise that performance is bad. Is it possible that your performance metrics aren't thorough enough? Is it just an engine limitation that is very hard to improve?

I feel like this major issue is being ignored. The flutter tag-line is "full native performance". In my few weeks testing I'd say frame rate is about 30% of native and it makes me sad because the platform has so much potential

**Comment by Chozzle** [comment]
I just read #813 and it seems the flutter team is prioritising this. Excellent!!!!

**Comment by liyuqian** [comment]
@Chozzle Thanks for your comments! Yes, performance is definitely our top priority too and we're working very hard on it 😄

**Comment by satish754ss** [comment]
After drawing lost of point paint become slow.
`` import 'dart:async';
import 'dart:typed_data';

import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:tahiti/activity_model.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter/rendering.dart';
import 'package:tahiti/popup_grid_view.dart';
import 'dart:ui' as ui show Codec, instantiateImageCodec, Image, FrameInfo;

class Drawing extends StatefulWidget {
  Drawing({
    this.template,
    Key key,
    this.model,
  }) : super(key: key);
  final ActivityModel model;
  Widget template;
  @override
  RollerState createState() {
    return new RollerState();
  }
}

class RollerState extends State<Drawing> {
  GlobalKey previewContainer = new GlobalKey();
  int count = 0;
  Offset pos;
  @override
  void initState() {
    super.initState();
  }

  @override
  void didUpdateWidget(Drawing oldWidget) {
    super.didUpdateWidget(oldWidget);
  }

  Drag _handleOnStart(Offset position) {
    widget.model.selectedThingId = '';
    widget.model.userTouch = false;
    if (count < 1) {
      setState(() {
        count++;
      });
      return _DragHandler(_handleDragUpdate, _handleDragEnd);
    }
    return null;
  }

  void _handleDragUpdate(DragUpdateDetails update) {
    PainterController painterController =
        ActivityModel.of(context).painterController;
    pos = (context.findRenderObject() as RenderBox)
        .globalToLocal(update.globalPosition);
    if (painterController.getDragStatus()) {
      painterController.updateCurrent(context, pos);
    } else {
      painterController.add(context, pos);
    }
  }

  void _handleDragEnd(DragEndDetails details) {
    ActivityModel model = ActivityModel.of(context);
    PainterController painterController = model.painterController;
    if (model.pathHistory.paths.length > 0) {
      painterController.endCurrent(context, pos);
      model.addDrawing(model.pathHistory.paths.last);
    }
    setState(() {
      count = 0;
    });
  }

  @override
  Widget build(BuildContext context) {
    return RepaintBoundary(
      key: previewContainer,
      child: LayoutBuilder(
        builder: (context, box) {
          return ClipRect(
              child: RawGestureDetector(
                  behavior: HitTestBehavior.opaque,
                  gestures: <Type, GestureRecognizerFactory>{
                    ImmediateMultiDragGestureRecognizer:
                        GestureRecognizerFactoryWithHandlers<
                            ImmediateMultiDragGestureRecognizer>(
                      () => ImmediateMultiDragGestureRecognizer(),
                      (ImmediateMultiDragGestureRecognizer instance) {
                        instance..onStart = _handleOnStart;
                      },
                    ),
                  },
                  child: _ScratchCardLayout(
                    child: Container(),
                    path: widget.model.pathHistory,
                    data: widget.model.painterController,
                  )));
        },
      ),
    );
  }
}

class _DragHandler extends Drag {
  _DragHandler(this.onUpdate, this.onEnd);

  final GestureDragUpdateCallback onUpdate;
  final GestureDragEndCallback onEnd;

  @override
  void update(DragUpdateDetails details) {
    onUpdate(details);
  }

  @override
  void end(DragEndDetails details) {
    onEnd(details);
  }
}

class _ScratchCardLayout extends SingleChildRenderObjectWidget {
  _ScratchCardLayout({
    Key key,
    this.path,
    this.strokeWidth = 25.0,
    @required this.data,
    @required this.child,
  }) : super(
          key: key,
          child: child,
        );

  final Widget child;
  final double strokeWidth;
  final PainterController data;
  final PathHistory path;

  @override
  RenderObject createRenderObject(BuildContext context) {
    return _ScratchCardRender(
        strokeWidth: strokeWidth,
        data: data,
        path: path,
        scratchCardContext: context);
  }

  @override
  void updateRenderObject(
      BuildContext context, _ScratchCardRender renderObject) {
    renderObject
      ..strokeWidth = strokeWidth
      ..data = data;
  }
}

class _ScratchCardRender extends RenderProxyBox {
  _ScratchCardRender({
    RenderBox child,
    double strokeWidth,
    this.path,
    this.scratchCardContext,
    PainterController data,
  })  : assert(data != null),
        _strokeWidth = strokeWidth,
        _data = data,
        super(child);

  double _strokeWidth;
  PainterController _data;
  final PathHistory path;
  BuildContext scratchCardContext;
  set strokeWidth(double strokeWidth) {
    assert(strokeWidth != null);
    if (_strokeWidth == strokeWidth) {
      return;
    }
    _strokeWidth = strokeWidth;
    markNeedsPaint();
  }

  set data(PainterController data) {
    assert(data != null);
    if (_data == data) {
      return;
    }
    if (attached) {
      _data.removeListener(markNeedsPaint);
      data.addListener(markNeedsPaint);
    }
    _data = data;

    markNeedsPaint();
  }

  @override
  void attach(PipelineOwner owner) {
    super.attach(owner);
    _data.addListener(markNeedsPaint);
  }

  @override
  void detach() {
    _data.removeListener(markNeedsPaint);
    super.detach();
  }

  PainterController painterController;

  @override
  void paint(PaintingContext context, Offset offset) {
    ActivityModel model = ActivityModel.of(scratchCardContext);
    if (child != null) {
      context.canvas.saveLayer(offset & size, Paint());
      context.paintChild(child, offset);
      if (model.painterController.drawingType == DrawingType.lineDrawing) {
        path.draw(context, size);
        path.drawStraightLine(context, size);
      } else {
        path.draw(context, size);
      }
      context.canvas.restore();
    }
  }

  @override
  bool get alwaysNeedsCompositing => child != null;
}``

**Comment by liyuqian** [comment]
@satish754ss Thanks for reporting the issue. I'm not sure whether it's related with the issue here. Maybe you can create a new issue with your sample code included so we can better track it.

**Comment by satish754ss** [comment]
Ok, thanks

**Comment by liyuqian** [comment]
https://github.com/flutter/flutter/issues/31086 is another example where it looks janky at 60fps.

**Comment by iapicca** [comment]
Hi @liyuqian 
Are you still experiencing the issue with the latest stable version of Flutter?
If so, can you please provide your `flutter doctor -v` 
and a reproducible minimal code sample?
Thank you

**Comment by jacob314** [comment]
Fyi @kenzieschmoll. It would be nice is the devtools timeline page could help with this case. For example, playback of the janky animation from screen captures at a lower frame rate to help the user understand what janky animation went on as well as just surfacing that the ui was janky in the timeline overview.

**Comment by Hixie** [comment]
I'm going to close this not because we shouldn't do it but because as described it's pretty vague. If we want to do this let's come up with a concrete design based on UXR and so forth and file a bug specifically for that work.
